File: Runtime/TransitionMultiplierValue.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports SyncroSim.StochasticTime

Friend Class TransitionMultiplierValue
    Inherits STSimDistributionBase

    Private m_TransitionGroupId As Integer
    Private m_StateClassId As Nullable(Of Integer)
    Private m_TransitionMultiplierTypeId As Nullable(Of Integer)

    Public Sub New(
        ByVal transitionGroupId As Integer,
        ByVal iteration As Nullable(Of Integer),
        ByVal timestep As Nullable(Of Integer),
        ByVal stratumId As Nullable(Of Integer),
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal stateClassId As Nullable(Of Integer),
        ByVal transitionMultiplierTypeId As Nullable(Of Integer),
        ByVal multiplierValue As Nullable(Of Double),
        ByVal distributionTypeId As Nullable(Of Integer),
        ByVal distributionFrequency As Nullable(Of DistributionFrequency),
        ByVal distributionSD As Nullable(Of Double),
        ByVal distributionMin As Nullable(Of Double),
        ByVal distributionMax As Nullable(Of Double))

        MyBase.New(
            iteration,
            timestep,
            stratumId,
            secondaryStratumId,
            multiplierValue,
            distributionTypeId,
            distributionFrequency,
            distributionSD,
            distributionMin,
            distributionMax)

        Me.m_TransitionGroupId = transitionGroupId
        Me.m_StateClassId = stateClassId
        Me.m_TransitionMultiplierTypeId = transitionMultiplierTypeId

    End Sub

    Public ReadOnly Property TransitionGroupId As Integer
        Get
            Return Me.m_TransitionGroupId
        End Get
    End Property

    Public ReadOnly Property StateClassId As Nullable(Of Integer)
        Get
            Return Me.m_StateClassId
        End Get
    End Property

    Public ReadOnly Property TransitionMultiplierTypeId As Nullable(Of Integer)
        Get
            Return Me.m_TransitionMultiplierTypeId
        End Get
    End Property

    Public Overrides Function Clone() As STSimDistributionBase

        Return New TransitionMultiplierValue(
            Me.TransitionGroupId,
            Me.Iteration,
            Me.Timestep,
            Me.StratumId,
            Me.SecondaryStratumId,
            Me.StateClassId,
            Me.TransitionMultiplierTypeId,
            Me.DistributionValue,
            Me.DistributionTypeId,
            Me.DistributionFrequency,
            Me.DistributionSD,
            Me.DistributionMin,
            Me.DistributionMax)

    End Function

End Class

File: src/Runtime/TransitionMultiplierValue.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using SyncroSim.StochasticTime;

namespace SyncroSim.STSim
{
    internal class TransitionMultiplierValue : STSimDistributionBase
    {
        private int m_TransitionGroupId;
        private int? m_StateClassId;
        private int m_AgeMin = 0;
        private int m_AgeMax = int.MaxValue;
        private int? m_TSTGroupId;
        private int m_TSTMin = 0;
        private int m_TSTMax = int.MaxValue;
        private int? m_TransitionMultiplierTypeId;

        public TransitionMultiplierValue(
            int transitionGroupId, 
            int? iteration, 
            int? timestep, 
            int? stratumId, 
            int? secondaryStratumId, 
            int? tertiaryStratumId, 
            int? stateClassId, 
            int ageMin,
            int ageMax,
            int? tstGroupId,
            int tstMin,
            int tstMax,
            int? transitionMultiplierTypeId, 
            double? multiplierValue, 
            int? distributionTypeId, 
            DistributionFrequency? distributionFrequency, 
            double? distributionSD, 
            double? distributionMin, 
            double? distributionMax) : base(
                iteration, timestep, stratumId, secondaryStratumId, 
                tertiaryStratumId, multiplierValue, distributionTypeId, 
                distributionFrequency, distributionSD, distributionMin, distributionMax)
        {

            this.m_TransitionGroupId = transitionGroupId;
            this.m_StateClassId = stateClassId;
            this.m_AgeMin = ageMin;
            this.m_AgeMax = ageMax;
            this.m_TSTGroupId = tstGroupId;
            this.m_TSTMin = tstMin;
            this.m_TSTMax = tstMax;
            this.m_TransitionMultiplierTypeId = transitionMultiplierTypeId;
        }

        public int TransitionGroupId
        {
            get
            {
                return this.m_TransitionGroupId;
            }
        }

        public int? StateClassId
        {
            get
            {
                return this.m_StateClassId;
            }
        }

        public int AgeMin
        {
            get
            {
                return this.m_AgeMin;
            }
        }

        public int AgeMax
        {
            get
            {
                return this.m_AgeMax;
            }
        }

        public int? TSTGroupId
        {
            get
            {
                return this.m_TSTGroupId;
            }
        }

        public int TSTMin
        {
            get
            {
                return this.m_TSTMin;
            }
        }

        public int TSTMax
        {
            get
            {
                return this.m_TSTMax;
            }
        }

        public int? TransitionMultiplierTypeId
        {
            get
            {
                return this.m_TransitionMultiplierTypeId;
            }
        }

        public override STSimDistributionBase Clone()
        {
            return new TransitionMultiplierValue(
                this.TransitionGroupId, 
                this.Iteration, 
                this.Timestep, 
                this.StratumId, 
                this.SecondaryStratumId, 
                this.TertiaryStratumId, 
                this.StateClassId, 
                this.AgeMin,
                this.AgeMax,
                this.TSTGroupId,
                this.TSTMin,
                this.TSTMax,
                this.TransitionMultiplierTypeId, 
                this.DistributionValue,
                this.DistributionTypeId, 
                this.DistributionFrequency, 
                this.DistributionSD, 
                this.DistributionMin, 
                this.DistributionMax);
        }
    }
}
